libraries/persistent_store/fuzz/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2019-2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO(ia0): Remove when used.
+#![allow(dead_code)]
+
+/// Bit-level entropy source based on a byte slice shared reference.
+struct Entropy<'a> {
+    /// The byte slice shared reference providing the entropy.
+    data: &'a [u8],
+
+    /// The bit position in the byte slice of the next entropy bit.
+    bit: usize,
+}
+
+impl Entropy<'_> {
+    /// Creates a bit-level entropy given a byte slice.
+    fn new(data: &[u8]) -> Entropy {
+        let bit = 0;
+        Entropy { data, bit }
+    }
+
+    /// Consumes the remaining entropy.
+    fn consume_all(&mut self) {
+        self.bit = 8 * self.data.len();
+    }
+
+    /// Returns whether there is entropy remaining.
+    fn is_empty(&self) -> bool {
+        assert!(self.bit <= 8 * self.data.len());
+        self.bit == 8 * self.data.len()
+    }
+
+    /// Reads a bit.
+    fn read_bit(&mut self) -> bool {
+        if self.is_empty() {
+            return false;
+        }
+        let b = self.bit;
+        self.bit += 1;
+        self.data[b / 8] & 1 << (b % 8) != 0
+    }
+
+    /// Reads a number with a given bit-width.
+    ///
+    /// # Preconditions
+    ///
+    /// - The number should fit in the return type: `n <= 8 * size_of::<usize>()`.
+    fn read_bits(&mut self, n: usize) -> usize {
+        assert!(n <= 8 * std::mem::size_of::<usize>());
+        let mut r = 0;
+        for i in 0..n {
+            r |= (self.read_bit() as usize) << i;
+        }
+        r
+    }
+
+    /// Reads a byte.
+    fn read_byte(&mut self) -> u8 {
+        self.read_bits(8) as u8
+    }
+
+    /// Reads a slice.
+    fn read_slice(&mut self, length: usize) -> Vec<u8> {
+        let mut result = Vec::with_capacity(length);
+        for _ in 0..length {
+            result.push(self.read_byte());
+        }
+        result
+    }
+
+    /// Reads a bounded number.
+    ///
+    /// # Preconditions
+    ///
+    /// - The bounds should be correctly ordered: `min <= max`.
+    /// - The upper-bound should not be too large: `max < usize::max_value()`.
+    fn read_range(&mut self, min: usize, max: usize) -> usize {
```

Comment:
Updated the documentation to stress that min and max are inclusive bounds (their mathematical sense).

This is not `random_int` since the returned values are constrained in a range, which is the word used in Rust to talk about an integer interval, in particular for random generation: https://rust-random.github.io/rand/rand/trait.Rng.html#method.gen_range

Reply:
Accepted for Rust random naming. (I was thinking of Python and numpy.)

The documentation still feels a bit light, since it sounds like it's a reasonably good random output. But if you'd compute a simple mean of your random numbers, you'd already converge to the wrong result. I'd mention that you don't intend to be a good random function, but you're just providing something in that particular range.

I'm pedantic on this, because bad RNG in a security application should really be something you're aware of.

Reply:
This is not an RNG which is why I call it entropy and I call the functions `read_` instead of `gen_`. I clarified the properties that matter in the documentation. It just has an interface that ressembles RNG but that's it. The usage is all but random, determinism is very important.

Reply:
True for not being an RNG. It's close enough that I was expecting RNG-like properties. And this is a stretch to me:

> the distribution is the closest to uniform for that entropy

Reply:
The proof is as follows:
- We have `n` bits coming from `ceil(log2(width))` as described in my other comments.
- We have to map `2^n` possibilities to `width` values.
- So the proba for each resulting value is of the form `X / 2^n` (which is obviously different from `1 / width` when `width != 2^n`)
- The best way to choose those X is to take `2^n - width` of them equal to 2 and the rest (2 * width - 2^n) equal to 1. Of course this is when `width < 2^n`. If `width = 2^n` then this is uniform by always taking since we have a bijection. Which ones we take as 2 or 1 doesn't matter.

Reply:
See discussion above for why this doesn't matter for the particular use case. Resolved.